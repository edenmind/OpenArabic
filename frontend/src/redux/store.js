import {
  arabicSentenceReducer,
  arabicTextReducer,
  arabicWordsReducer,
  authorReducer,
  categoryReducer,
  englishSentenceReducer,
  englishTextReducer,
  englishWordsReducer,
  sourceReducer,
  textReducer,
  wordByWordReducer,
} from './reducers'

import { configureStore } from '@reduxjs/toolkit'

export default configureStore({
  reducer: {
    title: textReducer,
    category: categoryReducer,
    author: authorReducer,
    source: sourceReducer,
    englishText: englishTextReducer,
    arabicText: arabicTextReducer,
    arabicWords: arabicWordsReducer,
    englishWords: englishWordsReducer,
    arabicSentence: arabicSentenceReducer,
    englishSentence: englishSentenceReducer,
    wordByWord: wordByWordReducer,
  },
})
