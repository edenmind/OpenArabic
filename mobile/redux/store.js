import { configureStore, combineReducers } from '@reduxjs/toolkit'

import { categoriesReducer, textReducer, textsReducer, UIStateReducer, wordsReducer } from './reducers.js'

const rootReducer = combineReducers({
  categories: categoriesReducer,
  text: textReducer,
  texts: textsReducer,
  textLoading: UIStateReducer,
  textsLoading: UIStateReducer,
  arabicFontSize: UIStateReducer,
  englishFontSize: UIStateReducer,
  isTransliterationOn: UIStateReducer,
  arabicFontName: UIStateReducer,
  practicingWords: UIStateReducer,
  words: wordsReducer,
  isDarkMode: UIStateReducer
})

const store = configureStore({
  reducer: rootReducer
})

export { store }
