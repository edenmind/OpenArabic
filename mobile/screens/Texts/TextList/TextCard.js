/* eslint-disable import/named */
/* eslint-disable import/namespace */
import { StyleSheet } from 'react-native';
import { Card, Chip, Avatar, Paragraph } from 'react-native-paper';
import React from 'react';
import * as utility from '../../../services/UtilityService';
import PropTypes from 'prop-types';

const style = StyleSheet.create({
  arabic: {
    direction: 'rtl',
    fontSize: 21,
    lineHeight: 30,
    writingDirection: 'rtl',
  },
  card: {
    marginBottom: 10,
    marginLeft: 10,
    marginRight: 10,
    marginTop: 10,
  },
  chip: {
    paddingBottom: 20,
    paddingLeft: 15,
    paddingTop: 20,
    color: '#a4cfbe',
  },
});

const LeftContent = (props) => (
  // eslint-disable-next-line react/jsx-props-no-spreading
  <Avatar.Icon {...props} icon="text" mode="elevated" />
);

export default function TextCard(props) {
  return (
    <Card style={style.card}>
      <Card.Cover source={{ uri: 'https://picsum.photos/700' }} />
      <Card.Title
        title={props.text.title}
        subtitle={`${props.text.author} (${props.text.timeAgo})`}
        left={LeftContent}
      />
      <Card.Content>
        <Paragraph>
          {utility.removeLineBreak(
            utility.truncate(`${props.text.englishText}`, 155),
          )}
        </Paragraph>
        <Paragraph style={style.arabic}>
          {utility.removeLineBreak(
            utility.truncate(`${props.text.arabicText}`, 125),
          )}
        </Paragraph>
      </Card.Content>
      <Card.Actions style={style.chip}>
        <Chip>{props.text.category}</Chip>
      </Card.Actions>
    </Card>
  );
}

TextCard.propTypes = {
  text: PropTypes.shape({
    title: PropTypes.string,
    author: PropTypes.string,
    timeAgo: PropTypes.string,
    englishText: PropTypes.string,
    arabicText: PropTypes.string,
    category: PropTypes.string,
  }),
};
