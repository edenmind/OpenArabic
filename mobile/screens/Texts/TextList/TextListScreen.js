/* eslint-disable import/named */
/* eslint-disable import/namespace */
/* eslint-disable react/forbid-prop-types */
import React, { useEffect, Fragment } from 'react';
import PropTypes from 'prop-types';
import { ScrollView, StyleSheet } from 'react-native';
import { Avatar, Button, Card, Paragraph, Divider } from 'react-native-paper';
import * as utility from '../../../services/UtilityService';
import * as api from '../../../services/ApiService';
import { useDispatch, useSelector } from 'react-redux';
import Spinner from '../../../components/Spinner';

export function TextListScreen({ route, navigation }) {
  const { category } = route.params;
  const SHARE = 'Share';
  const LIKE = 'Like';

  const { texts } = useSelector((state) => state.texts);
  const dispatch = useDispatch();
  const fetchTexts = () => dispatch(api.getTexts(category, 7, 0));

  useEffect(() => {
    fetchTexts();
  });

  const style = StyleSheet.create({
    arabic: {
      direction: 'rtl',
      fontSize: 21,
      lineHeight: 30,
      writingDirection: 'rtl',
    },
    card: {
      marginBottom: 5,
      marginLeft: 10,
      marginRight: 10,
      marginTop: 5,
    },
  });

  const LeftContent = (props) => (
    // eslint-disable-next-line react/jsx-props-no-spreading
    <Avatar.Icon {...props} icon="text" mode="elevated" />
  );

  const textItems = texts.map((text) => (
    <Fragment key={text.textId}>
      <Card
        style={style.card}
        onPress={() =>
          navigation.navigate('SingleText', {
            textId: text.textId,
          })
        }>
        <Card.Cover source={{ uri: 'https://picsum.photos/700' }} />
        <Card.Title
          title={text.title}
          subtitle={`${text.author} (${text.readTime})`}
          left={LeftContent}
        />
        <Card.Content>
          <Paragraph>
            {utility.removeLineBreak(
              utility.truncate(`${text.englishText}`, 155),
            )}
          </Paragraph>
          <Paragraph style={style.arabic}>
            {utility.removeLineBreak(
              utility.truncate(`${text.arabicText}`, 125),
            )}
          </Paragraph>
        </Card.Content>
        <Card.Actions>
          <Button>{LIKE}</Button>
          <Button>{SHARE}</Button>
        </Card.Actions>
      </Card>
      <Divider />
    </Fragment>
  ));

  if (texts.length > 1) {
    return <ScrollView>{textItems}</ScrollView>;
  }
  return <Spinner />;
}

TextListScreen.propTypes = {
  route: PropTypes.any.isRequired,
  navigation: PropTypes.any.isRequired,
};
