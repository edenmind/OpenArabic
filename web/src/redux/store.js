import * as reducers from './reducers'

import { configureStore } from '@reduxjs/toolkit'

export default configureStore({
  reducer: {
    text: reducers.textReducer,
  },
})
