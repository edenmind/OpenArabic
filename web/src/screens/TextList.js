import * as apiService from '../services/apiService'

import { Box, Grid } from '@mui/material'
import React, { Fragment } from 'react'

import Progress from '../components/Progress'
import PropTypes from 'prop-types'
import TextListCard from './TextListCard'

const TextList = (props) => {
  const [texts, setTexts] = React.useState([])
  const [isLoading, setIsLoading] = React.useState(true)

  React.useEffect(() => {
    apiService
      .getTexts(props.id)
      .then((data) => {
        setTexts(data)
        setIsLoading(false)
      })
      .catch((err) => console.log(err))
  }, [props.id])

  return isLoading ? (
    <Progress />
  ) : (
    <Fragment>
      <h2>{props.heading}</h2>
      <h4>{props.subHeading}</h4>
      <Box sx={{ flexGrow: 1 }}>
        <Grid container spacing={2}>
          <TextListCard texts={texts} />
        </Grid>
      </Box>
    </Fragment>
  )
}

TextList.propTypes = {
  heading: PropTypes.node,
  subHeading: PropTypes.node,
  id: PropTypes.string,
}

export default TextList
