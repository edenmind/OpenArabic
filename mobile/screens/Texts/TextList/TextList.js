/* eslint-disable import/named */
/* eslint-disable import/namespace */
/* eslint-disable react/forbid-prop-types */
import React, { useEffect, useState } from 'react';
import PropTypes from 'prop-types';
import { Pressable, ScrollView } from 'react-native';
import * as api from '../../../services/ApiService';
import { useDispatch, useSelector } from 'react-redux';
import TextCard from './TextCard';
import Spinner from '../../../components/Spinner';
import { useFocusEffect } from '@react-navigation/native';

export function TextList({ route, navigation }) {
  const { category } = route.params;
  const [isLoading, setIsLoading] = useState(true);
  const [shouldReload, setShouldReload] = useState(true);
  const selector = (state) => state.texts;
  const { texts } = useSelector(selector);
  const dispatch = useDispatch();
  const fetchTexts = () => {
    if (category == 'Home') {
      dispatch(api.getTexts('', 25, 0));
    } else {
      dispatch(api.getTexts(category, 50, 0));
    }

    setTimeout(() => {
      setIsLoading(false);
    }, 1500);
  };

  useFocusEffect(
    React.useCallback(() => {
      if (shouldReload) {
        setIsLoading(true);
        fetchTexts();
      }
    }, [shouldReload]),
  );

  const cards = texts.map((text) => (
    <Pressable
      key={text.textId}
      onPress={() => {
        navigation.navigate('TextScreen', {
          textId: text.textId,
        });
        setShouldReload(false);
      }}>
      <TextCard text={text}></TextCard>
    </Pressable>
  ));

  if (isLoading) {
    return <Spinner />;
  } else {
    return <ScrollView>{cards}</ScrollView>;
  }
}

TextList.propTypes = {
  route: PropTypes.any.isRequired,
  navigation: PropTypes.any.isRequired,
};
