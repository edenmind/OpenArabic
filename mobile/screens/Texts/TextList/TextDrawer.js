import React, { useState, useEffect } from "react";
import { StyleSheet } from "react-native";
import { TextListScreen } from "./TextListScreen";
import { createDrawerNavigator } from "@react-navigation/drawer";
import axios from "axios";

const Drawer = createDrawerNavigator();

export function TextDrawer() {
  const [categories, setCategories] = useState([{ categoryId: 0, name: "Adab" }]);
  const categoryItems = categories.map((category) => (
    <Drawer.Screen name={category.name} component={TextListScreen} initialParams={{ category: category.name }} options={{ title: category.name }} key={category.name} />
  ));

  useEffect(() => {
    axios.get("https://api.openarabic.io/api/categories").then(function (response) {
      setCategories(response.data);
    });
  }, [categoryItems]);

  return <Drawer.Navigator initialRouteName="Texts">{categoryItems}</Drawer.Navigator>;
}

export const style = StyleSheet.create({
  container: {
    flex: 1,
    alignItems: "center",
    justifyContent: "center",
  },
});
