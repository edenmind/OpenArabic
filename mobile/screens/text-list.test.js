/* eslint-disable putout/objects-braces-inside-array */
import { describe, it, jest } from '@jest/globals'
import TextList from './text-list.js'
import { NavigationContainer } from '@react-navigation/native'
import { Provider } from 'react-redux'
import React from 'react'
import ShallowRenderer from 'react-test-renderer/shallow' // ES6
import { store } from '../redux/store.js'

jest.useFakeTimers()
jest.mock('react-native/Libraries/Animated/NativeAnimatedHelper')

jest.mock('react-redux', () => ({
  ...jest.requireActual('react-redux'),
  useSelector: () => ({
    text: {
      textLoading: true,
      text: {
        arabic: ['a1', 'a2', 'a3'],
        english: ['e1', 'e2', 'e3'],
        status: 'Draft',
        title: 'abc',
        publishAt: '2021-01-01T00:00:00.000Z',
        wordByWord: [['']],
        image: 'abc',
        texts: { arabic: '', english: '' },
        category: 'abc',
        author: '',
        arabicSentence: [''],
        source: '',
        vocabularyCollection: {
          numberOfBatches: 1,
          arabic: [[{ word: 'a', wordId: '1' }]],
          english: [[{ word: 'e', wordId: '1' }]]
        },
        // eslint-disable-next-line putout/objects-braces-inside-array
        sentences: [
          {
            english: 'e1',
            arabic: 'a1',
            words: [
              {
                arabic: 'a',
                english: '1'
              }
            ]
          },
          {
            english: 'e2',
            arabic: 'a2',
            words: [
              {
                arabic: 'a',
                english: '2'
              }
            ]
          },
          {
            english: 'e3',
            arabic: 'a3',
            words: [
              {
                arabic: 'a',
                english: '3'
              }
            ]
          }
        ]
      }
    }
  })
}))

describe('<Category />', () => {
  it('renders correctly', () => {
    const renderer = new ShallowRenderer()
    const tree = renderer.render(
      <Provider store={store}>
        <NavigationContainer>
          <TextList route={'All'} />
        </NavigationContainer>
      </Provider>
    )

    expect(tree).toMatchSnapshot()
  })
})
