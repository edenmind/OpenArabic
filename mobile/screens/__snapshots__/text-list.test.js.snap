// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Category /> renders correctly 1`] = `
<Context.Provider
  value={
    {
      "getServerState": undefined,
      "store": {
        "@@observable": [Function],
        "dispatch": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "subscription": {
        "addNestedSub": [Function],
        "getListeners": [Function],
        "handleChangeWrapper": [Function],
        "isSubscribed": [Function],
        "notifyNestedSubs": [Function],
        "trySubscribe": [Function],
        "tryUnsubscribe": [Function],
      },
    }
  }
>
  <ForwardRef(NavigationContainerInner)>
    <TextList />
  </ForwardRef(NavigationContainerInner)>
</Context.Provider>
`;
